### ISSUE TITLE
^goto if does not emit size optimal code

### ISSUE THREAD

**Issue by undisbeliever** [wiz-lang/wiz#64]
A `^goto if` emits non size optimal code on:

synthetic comparison (see [6502_goto_far_synthetic.wiz](https://github.com/undisbeliever/wiz/blob/test-suite/tests/block/6502_goto_far_synthetic.wiz))
```
    ^goto far_label if a <= 1;
``` 
which emits:
```
// 000000      c9 01                 cmp #0x01
//             d0 03                 bne 0x008007
//             4c ed fe              jmp 0xfeed
// 000007      b0 03                 bcs 0x00800c
//             4c ed fe              jmp 0xfeed
```

when a shorter (and just as fast) output should be:
```
    cmp     #1
    beq     +
    bcs     ++
    +
        jmp     0x8000
    ++
```
<br/>

The same goes for `>` comparison
```
// BLOCK 000048      c9 01                 cmp #0x01
// BLOCK             d0 03                 bne 0x00804f
// BLOCK             4c 54 80              jmp 0x8054
// BLOCK 00004f      90 03                 bcc 0x008054
// BLOCK             4c ed fe              jmp 0xfeed
    ^goto far_label if a > 1;
```

Which should be compiled to:
```
    cmp #1
    beq +
    bcc +
        jmp 0x8000
    +
```
<br/>

This bug also affects `^goto if` statements involving logical operators (see [6502_goto_and_or.wiz](https://github.com/undisbeliever/wiz/blob/709673065a7b5cf9db0e02361dee74576e4b781f/tests/block/6502_goto_and_or.wiz#L130) for more examples)
```
// BLOCK 0000e7      c9 ab                 cmp #0xab
// BLOCK             d0 03                 bne 0x0080ee
// BLOCK             4c ed fe              jmp 0xfeed
// BLOCK 0000ee      e0 cd                 cpx #0xcd
// BLOCK             f0 03                 beq 0x0080f5
// BLOCK             4c fc 80              jmp 0x80fc
// BLOCK 0000f5      c0 ef                 cpy #0xef
// BLOCK             d0 03                 bne 0x0080fc
// BLOCK             4c ed fe              jmp 0xfeed
    ^goto FarLabel if
        a == 0xab
        || x == 0xcd
        && y == 0xef;
```

**Comment by Bananattack** [comment]
I think the issue here is that the `^` indicator currently applies to each sub-branch rather than taking advantage of the overall branching already being 'far' away. Hopefully this system could be improved.

**Comment by undisbeliever** [comment]
Yeah, I think a simple solution for the bug would be to convert `^goto if <condition>` statements into something like:
```
if <condition> {
    ^goto label;
}
```

**Comment by Bananattack** [comment]
So we summarized the problem above, but at its root, we need to generate more optimal code for absolute jumps when there are logical and/or operators. And I think it's worth working through some very basic examples to arrive at the exact code-gen we want out of this.

Current code-gen for `||` (when there is no conditional absolute jump):

```
^goto label if a || b;
// same as:
^goto label if a;
^goto label if b;
// same as:
goto skip if !a;
^goto label;
skip:
goto skip2 if !a;
^goto label;
skip2:
```

This is suboptimal, because there's an extra absolute branch. What we want for `||` is this instead:

```
goto success if a;
goto skip if !b;
success:
^goto label;
skip:
```

Current code-gen for `&&` (when there is no conditional absolute jump):

```
^goto label if a && b;
// same as:
goto skip if !a;
^goto label if b;
skip:
// same as:
goto skip if !a;
goto skip2 if !b;
^goto label;
skip2:
skip:
```

Since skip2 and skip will be labels that map to the same address, it looks like non-negated `&&` might actually already be doing the right thing?

However, `!(a && b)` is probably going to have the same truth short-circuit optimization case as `a || b`, due to the part of De Morgan's Laws which state that `!(a && b) = (!a || !b)`.

I think comparison operators `>` and `<=` can be thought of as `!(less || equal) = !less && !equal` and `less || equal` for the sake of how their code should be generated. And these will suffer the same issue for `<=` because of the suboptimal or-chain generation.

**Comment by Bananattack** [comment]
Hmm so investigating this further, it may be tricky because currently the synthesized long-branches are done as low-level instruction matches. A pattern that describes "distance hint >= 1, zero flag, branch if true" becomes the raw opcode sequence for "bne skip, jmp label"

Since this done at the raw instruction substitution level, the two sub-branches are an inseparable opaque blob to the compiler. For this to be fixed, we need to remove these synthetic instructions from the 6502, etc backends, and add higher level support for branch synthesis that first attempts and fails to find an absolute conditional branch, and then falls back on synthesis of a relative conditional branch-on-opposite and an unconditional absolute jump. Then in that case, we'd be able to inject a "success" branch in or-chains.

Not impossible but a sizable undertaking potentially. Maybe there's an easier way that can catch this before going to instruction-matching though, and decompose to branch-on-opposite-then jump. Hmm.

I mean, it'd be useful for even platforms that have absolute conditional branches, since only the last branch in the chain needs to be absolute, the rest can be relative to save space and potentially cycles. Anyways, it seems this has some tricky bits, may need to spin a while longer to get a good answer to this
